riscv : add basic single precision hw float support

mean00 opened the pull request:
## Detailed description

* This adds basic support for hw single precision fpu (rv32)
* Without the patch gdb will reject the binary as the register list is incomplete
* This is not exactly the same as the patch posted in the bug report, it actually connects the registers
* With that patch on a CH32V307 , i can load the binary and print the registers with a value that looks correct (limited testing), seems ok for fs0, f1, fs2 , fs3
* Tested only on rv32
* May lack some CSR relatated to FPU declared as CSR , not clear.



* [X] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [X] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [up to linking then out of rom ] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware))
* [X] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [X] I've tested it to the best of my ability
* [X] My commit messages provide a useful short description of what the commits do

## Closing issues
fixes #1765

dragonmux commented on src/target/riscv32.c:
```diff
@@ -153,6 +153,9 @@ static ssize_t riscv32_reg_read(target_s *target, const uint32_t reg, void *data
 		return riscv32_bool_to_4(riscv_csr_read(hart, RV_DPC, data));
 	if (reg >= RV_CSR_GDB_OFFSET)
 		return riscv32_bool_to_4(riscv_csr_read(hart, reg - RV_CSR_GDB_OFFSET, data));
+	if(reg >= RV_FPU_GDB_FIRST_REGISTER) {
```
Please drop the additional braces in this and the register write if statement additions - they're not necessary. To be more consistent with the style of the existing code and how things are done in the other architecture implementations, may we suggest `RV_FPU_GDB_OFFSET` be used for this constant?

Please also run clang-format - there's an extra space in the second change here, and this + the other added if statement are missing one between `if` and the `(` for the condition.

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -157,6 +157,15 @@ static const riscv_csr_descriptor_s riscv_csrs[]={
+										"s0","s1","a0","a1","a2","a3","a4",
+										"a5","a6","a7","s2","s3","s4","s5",
+										"s6","s7","s8","s9","s10","s11","t8",
+										"t9","t10","t11"};
```
The final string needs to be followed by a `,` to get clang-format to format this properly.

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -999,7 +1047,34 @@ static size_t riscv_build_target_description(
+	
+
+	/* Basic single precision support */
+	if (extensions & RV_ISA_EXT_SINGLE_FLOAT) {
```
Given this needs to essentially be done again for doubles, please move this into its own function above this one. That way it can be re-used more easily to build out the `RV_ISA_EXT_DOUBLE_FLOAT` handling.

dragonmux commented on src/target/riscv_debug.h:
```diff
@@ -204,6 +204,8 @@ typedef struct riscv_hart {
 // so they cannot collide with GPR registers, so you have to subtract
 // RV_CSR_GDB_OFFSET from the value received from gdb
 #define RV_CSR_GDB_OFFSET 128
+#define RV_FPU_GDB_FIRST_REGISTER 33
```
The original idea with this section of defines is that it was a block just for the CSRs - perhaps move these two new defines below to line 218 as a new block just for FPU definitions?

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -999,7 +1047,34 @@ static size_t riscv_build_target_description(
+			if (max_length != 0)
+				print_size = max_length - (size_t)offset;
+			offset += snprintf(buffer + offset, print_size,
+				"<reg name=\"f%s\" bitsize=\"32\" type=\"ieee_single\" regnum=\"%u\" group=\"float\"/>",
```
One notable thing from our prior work with GDB here, we can likely drop both the `group=` and `type=` stanzas which reduces the size of the feature blob generated here, which can otherwise get rather large. This reduces the likelihood of an OOM condition occurring on STM32F1-based probes.

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -999,7 +1047,34 @@ static size_t riscv_build_target_description(
+			print_size = max_length - (size_t)offset;
+		offset += snprintf(buffer + offset, print_size, "</feature><feature name=\"org.gnu.gdb.riscv.fpu\">");
+
+		for (unsigned int i = 0; i < sizeof(riscv_fpu_regs) / sizeof(riscv_fpu_regs[0]); i++) {
```
Please use the `ARRAY_LENGTH()` macro rather than computing the array length yourself here. Likewise `size_t` not `unsigned int` - the type of sizeof() is always size_t, not using this type alias may lead to type conversions.

dragonmux reviewed (commented):
This is looking good though needs `clang-format` run - please see our review notes below for the stylistic items we've spotted and some code organisation suggestions.

With those taken care of, this should be good to merge.

mean00 commented:
Thank you
I've fixed most if not all as a separate commit. If it's okay i'll squash them

dragonmux commented:
It's looking much better. Please make sure you run `clang-format` on the changes as there are still inconsistencies that it will take care of, and that leaves just one review item to discuss and address.

mean00 commented:
I have both git pre commit installed & manually ran clang-format before commiting (twice).
 I'm beginning to suspect having clang-fmt 17 is causing the issues. It looks like bmp is setup for clang-fmt16.  Maybe a known gotcha ?

Regarding extracting the function, i would have to forward fpu size (32/64) but also offset, buffer,  .... as parameters
Am i missing something or is it the idea ?

Thank you

dragonmux commented:
clang-format-17 should work fine (it's clang-format-18 that does extra stuff that's unwanted here), but failing that use clang-format-16 via `pip` (https://github.com/ssciwr/clang-format-wheel)

Re extracting the function: yes, that's exactly the idea - see https://github.com/blackmagic-debug/blackmagic/blob/c4015cdf5efaeaca35075f04f461d72132aa3a37/src/target/cortexar.c#L1682-L1704 and the Cortex-M equivalent which explicitly and intentionally splits out FPU block generation to manage routine complexity and size. Ideally the CSR block would likewise be split out.

mean00 commented:
Thank you for the details
Hopefully this time it is ok.
I've moved the fpu part to its own function (like for CM-AR) and added a fpu size parameter so it should be easier to add support for double precision later (dont have any of those available to test).

(triple checked clang-format (17) was not complaining, not too hopeful)

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -965,6 +1004,45 @@ static const char *riscv_fpu_ext_string(const uint32_t extensions)
  * Both are then continued with:
  * "		<reg name=\"pc\" bitsize=\"[address_width]\" type=\"code_ptr\"/>"
  * "	</feature>"
+ * The following are only generated for an F core (single precision HW float support):
```
We would move this comment block to the front of the FPU block function so the documentation for it is kept together with the function.

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -918,6 +927,36 @@ static const char *riscv_fpu_ext_string(const uint32_t extensions)
+ * fpu_size =32 single precision float
+ * fpu_size =64 double precision float
+ */
+static size_t riscv_build_target_fpu_description(size_t fpu_size, char *const buffer, size_t max_length)
```
Given the order of parameters on the base function, the compiler will likely generate a smaller amount of code for the call frame if `fpu_size` is made the last parameter. This is fine otherwise.

dragonmux commented on src/target/riscv_debug.c:
```diff
@@ -918,6 +927,36 @@ static const char *riscv_fpu_ext_string(const uint32_t extensions)
+static size_t riscv_build_target_fpu_description(size_t fpu_size, char *const buffer, size_t max_length)
+{
+	const size_t first_fpu_register = RV_FPU_GDB_OFFSET;             // fixme!
+	const size_t first_fpu_control_register = RV_FPU_GDB_CSR_OFFSET; // fixme!
```
It would be useful with this and the line above if you could expand on what needs fixing about them - it's very curious what those "fixme!" comments are about.

mean00 replied:
No sure i got it correctly, i think the idea was to move the clang-format on up
+ fixed a mistake on fpu csr param that were swapped
Thank you

dragonmux replied:
Think you'd meant to reply to the first of these 4 review items. Yes, that is the change we were thinking though you've moved it further up than we were anticipating - but think that's ok as that other code needed some formatting love too.

The only remaining issue there is the lack of a blank line between the preceding array and your addition, but as the CSR stuff needs some work, don't worry about that. We can fix that after.

dragonmux reviewed (commented):
Think we've finally figured out why clang-format was not formatting those FPU regs strings - see our review notes. This is almost there. There's a query we've got on two lines in the FPU block generator and a couple of suggestions and once those are answered and taken care of w/ the one necessary fix made, this should be good to merge.

dragonmux reviewed (approved):
This LGTM now, we'll get this merged as we have no reasonable way to test it locally.

mean00 commented:
Squashed, thank you
